File: sortlinkedlist.py
'''
How to sort a linked list in Python
This code implements a merge sort algorithm to sort a linked list.
This is a common interview question and can be solved using a recursive approach.
This code defines a `Node` class for the linked list and a `LinkedList` class
that contains methods to insert nodes, merge two sorted lists, and sort the linked list using merge sort.
'''
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        
class LinkedList:   
    def __init__(self):
        self.head = None

    def insert(self, data):
        new_node = Node(data)
        if not self.head:
            self.head = new_node
            return
        last = self.head
        while last.next:
            last = last.next
        last.next = new_node

    def merge(self, left, right):
        if not left:
            return right
        if not right:
            return left
        if left.data < right.data:
            result = left
            result.next = self.merge(left.next, right)
        else:
            result = right
            result.next = self.merge(left, right.next)
        return result

    def get_middle(self, head):
        if not head or not head.next:
            return head
        slow = head
        fast = head.next
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        return slow

    def merge_sort(self, head):
        if not head or not head.next:
            return head
        mid = self.get_middle(head)
        next_to_mid = mid.next
        mid.next = None  # Split the list into two halves

        left_half = self.merge_sort(head)
        right_half = self.merge_sort(next_to_mid)

        return self.merge(left_half, right_half)

    def sort(self):
        self.head = self.merge_sort(self.head)

    def print_list(self):
        current = self.head
        while current:
            print(current.data, end=' ')
            current = current.next
    def __str__(self):
        result = []
        current = self.head
        while current:
            result.append(str(current.data))
            current = current.next
        return ' -> '.join(result)

if __name__ == "__main__":
    ll = LinkedList()
    ll.insert(4)
    ll.insert(2)
    ll.insert(1)
    ll.insert(3)
    
    print("Original linked list:")
    ll.print_list()
    
    ll.sort()
    
    print("\nSorted linked list:")
    ll.print_list()

File: sortlinkedlist.cpp
#include <iostream>
using namespace std;

struct Node {
    int data;
    Node* next;
    Node(int d): data(d), next(nullptr) {}
};

// Utility: print list
void printList(Node* head) {
    for (auto p = head; p; p = p->next)
        cout << p->data << " ";
    cout << "\n";
}

// Split the list into two halves, return middle node
Node* getMiddle(Node* head) {
    if (!head) return head;
    Node *slow = head, *fast = head->next;
    while (fast && fast->next) {
        slow = slow->next;
        fast = fast->next->next;
    }
    return slow;
}

// Merge two sorted lists
Node* mergeLists(Node* a, Node* b) {
    if (!a) return b;
    if (!b) return a;
    if (a->data < b->data) {
        a->next = mergeLists(a->next, b);
        return a;
    } else {
        b->next = mergeLists(a, b->next);
        return b;
    }
}

// Recursive merge-sort
Node* mergeSort(Node* head) {
    if (!head || !head->next) return head;
    Node* mid = getMiddle(head);
    Node* right = mid->next;
    mid->next = nullptr;
    Node* left_sorted  = mergeSort(head);
    Node* right_sorted = mergeSort(right);
    return mergeLists(left_sorted, right_sorted);
}

int main() {
    // build list: 4→2→1→3
    int vals[] = {4, 2, 1, 3};
    Node* head = new Node(vals[0]);
    Node* p = head;
    for (int i = 1; i < 4; ++i) {
        p->next = new Node(vals[i]);
        p = p->next;
    }

    cout << "Before: ";
    printList(head);

    head = mergeSort(head);

    cout << "After:  ";
    printList(head);
    return 0;
}

//g++ -std=c++11 linked_list_mergesort.cpp -o linked_list_mergesort
//./linked_list_mergesort